### PR TITLE
ENH: drop dtype metadata

### LINKED ISSUE #23169: BUG: unable to read .npy file with non-eval'able metadata
### Describe the issue:

I have a .npy file generated from an h5py dataset. The data is a pure Numpy array (no subclass) and the file has been written by Numpy, but the dtype contains metadata set by h5py.
In this particular case, the metadata contains a `<class 'bytes'>` expression, which cannot be handled by `safe_eval` . I don't care about the metadata, but it renders the entire file unreadable.

```text
 File "/home/sjoerd/miniconda3/envs/active-papers/lib/python3.9/site-packages/numpy/lib/format.py", line 776, in read_array
    shape, fortran_order, dtype = _read_array_header(
  File "/home/sjoerd/miniconda3/envs/active-papers/lib/python3.9/site-packages/numpy/lib/format.py", line 632, in _read_array_header
    raise ValueError(msg.format(header)) from e
ValueError: Cannot parse header: "{'descr': [('paper_ref', ('|O', {'vlen': <class 'bytes'>})), ('path', ('|O', {'vlen': <class 'bytes'>}))], 'fortran_order': False, 'shape': (), }                                    \n"
```

I  would consider this a bug in `np.save` . Serialization of a dtype descriptor that results in an unreadable header should fail loudly. It would be nice if the `np.save` could then be re-tried with the metadata stripped.

### Reproduce the code example:

```python
import numpy as np
np.load("file.npy")
```

### Error message:

```shell
Traceback (most recent call last):
  File "/home/sjoerd/bayesian_inference_fm/aptool-dump.py", line 65, in dump
    arr = np.load(child_target + ".npy")
  File "/home/sjoerd/miniconda3/envs/active-papers/lib/python3.9/site-packages/numpy/lib/npyio.py", line 432, in load
    return format.read_array(fid, allow_pickle=allow_pickle,
  File "/home/sjoerd/miniconda3/envs/active-papers/lib/python3.9/site-packages/numpy/lib/format.py", line 776, in read_array
    shape, fortran_order, dtype = _read_array_header(
  File "/home/sjoerd/miniconda3/envs/active-papers/lib/python3.9/site-packages/numpy/lib/format.py", line 632, in _read_array_header
    raise ValueError(msg.format(header)) from e
ValueError: Cannot parse header: "{'descr': [('paper_ref', ('|O', {'vlen': <class 'bytes'>})), ('path', ('|O', {'vlen': <class 'bytes'>}))], 'fortran_order': False, 'shape': (), }                                    \n"
```

### Runtime information:

Numpy version:

1.24.0
3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:39:03) 
[GCC 11.3.0]

show_runtime:

WARNING: `threadpoolctl` not found in system! Install it by `pip install threadpoolctl`. Once installed, try `np.show_runtime` again for more detailed build information
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]

### Context for the issue:

A pure Numpy array (no subclasses) saved with `np.save` should be readable by `np.load`. This issue gives a counter-example.

I could try to write a patch to fix this issue, if you are interested.

### PR DESCRIPTION
**Pull request by ninousf**
the continuation of the  PR #23185

(Doesn't actually fix the `descr` part including things, but the important thing here)
Closes gh-23169, gh-15488

### CODE REVIEW

File: numpy/lib/utils.py [hunk 1]
```diff
@@ -1140,4 +1140,31 @@ def _opt_info():
...
+        # structured dtype
+        l = list()
+        for n in dt.names:
+            t = dt.fields[n]
+            l.append((
+                n,  # name
+                drop_metadata(t[0]),  # format
+                t[1],  # offset
+                t[2] if len(t) == 3 else None,  # title
+            ))
```
Comment (MatteoRaso) [inline numpy/lib/utils.py:1156]:
```suggestion
            if len(t) == 3:
                l.append((n, drop_metadata(t[0]), t[1], t[2]))

            else:
                l.append((n, drop_metadata(t[0]), t[1], None))
```
I think this is much more readable.

File: numpy/lib/utils.py [hunk 2]
```diff
@@ -1140,4 +1140,31 @@ def _opt_info():
             enabled_features += f" {feature}?"
 
     return enabled_features
+
+def drop_metadata(dt):
+    align = dt.isalignedstruct
```
Comment (MatteoRaso) [inline numpy/lib/utils.py:1145]:
```suggestion
```
This variable is unused, so it should be removed.

**Review by MatteoRaso — COMMENTED** [review]
Thanks for the PR. It looks great, just a few small suggestions.

**Comment by seberg** [comment]
@ninousf sorry for not following up earlier?  Did you have a use-case beyond dropping metadata in `.npy`?
Otherwise, I am tempted to follow-up to use this here and consider this semi-private (which it effectively is).

**Comment by ninousf** [comment]
Hi Sebastian
We can keep it in utils.py
My use case is to be able to copy a hdf5 file which contains arrays with dtypes containing metadata in a zarr file
See #1152

**Comment by seberg** [comment]
It looks a bit like we may add `np.dtypes` as a namespace (I am starting to assume we will have enough use for it).  That is a bit annoying, because adding it here might be too much right now, so I can follow-up on it.

One thing I am wondering:  I somewhat like the idea of adding the behavior:
* Check whether `metadata is None` ahead of time, if it is return the dtype unmodified (the identity).
* For a structured/subarray dtypes we would have to also return the identity if all fields/the subdtype and the dtype itself has no metadata.

That would allow us to do:
```
no_metadata = drop_metadata(dtype)
if no_metadata is not dtype:
    warnings.warn("Dropping metadata which cannot be stored in an npy/npz file", UserWarning)
```
Although, I am not sure we actually need the warning there, maybe it is useful to be able to know that a change happened?

**Comment by seberg** [comment]
@ninousf could you review my changes?  I made some heavier changes to use this for `np.save` and `np.savez` and also just to make the code cleaner, IMO.

Now, I don't *like* `np.lib.utils.drop_metadata` as a home, so I added a fuzzy "this is somewhat public" comment, but :shrug:.
I would have liked to put this into the new `np.dtypes` namespace.  But the function really has some problems (poor record dtype and results are plain wrong for user dtypes).  That fits OK with npy saving, which has the same problem, but makes me not want to put it there yet.  We should find a proper solution (probably in C).

The new code is a bit more robust than the previous one, because it returns the original dtype when there is no metadata attached.  So at least if no metadata is used things should be fine even for user dtypes.

**Comment by ninousf** [comment]
@seberg it's ok for me, thanks

**Comment by charris** [comment]
This could use a release note.

**Comment by charris** [comment]
@ninousf This needs a release note. If you cannot add that today, I'm going to push this off to the next release.

**Comment by ninousf** [comment]
@charris Tell me if it's ok for you

**Comment by charris** [comment]
Looks fine, thanks.

**Comment by charris** [comment]
Thanks @ninousf

**Comment by ntessore** [comment]
Please consider marking this as fully public. Metadata is great to forego ndarray subclasses, but until this function came along, there was no good way to e.g. update a value in the metadata.

**Pull request merged**